dwave/cloud/solver.py

```diff
@@ -1005,13 +1008,11 @@ def check_problem(self, linear, quadratic):
             False
             True
         """
-        for key, value in uniform_iterator(linear):
-            if value != 0 and key not in self.nodes:
-                return False
-        for key, value in uniform_iterator(quadratic):
-            if value != 0 and tuple(key) not in self.edges:
-                return False
-        return True
+        # handle legacy format
+        if not isinstance(linear, Mapping):
+             linear = {idx: val for idx, val in enumerate(linear) if val != 0}
```

Comment:
It doesn't really matter, but since you're just checking the keys of `linear` you could do
```python
linear = [idx for idx, val in enumerate(linear) if val != 0]
```
or even use [np.nonzero](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html).

Reply:
I went with a dict because it's already set-like for keys, and `issuperset(linear)` doesn't have to build a set. Not sure it actually make any difference, especially for a small list of linear terms...

Reply:
Ah, interesting! It does indeed seem like they are casting to a set under the hood (at least based on timings)
```python
In [1]: a = set(range(1000))

In [2]: b_list = list(range(500, 1000))

In [3]: b_set = set(range(500, 1000))

In [4]: b_dict = {v: 0 for v in range(500, 1000)}

In [5]: %timeit a.issuperset(b_list)
13.2 µs ± 68.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit a.issuperset(set(b_list))
13.4 µs ± 183 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit a.issuperset(b_set)
7.25 µs ± 21.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit a.issuperset(b_dict)
11.6 µs ± 48.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [9]: %timeit a.issuperset(set(b_dict))
11.8 µs ± 69.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
which kind of surprises me. I would have expected that any iterable would be fine.

Reply:
And so they do https://github.com/python/cpython/blob/f25f2e2e8c8e48490d22b0cdf67f575608701f6f/Objects/setobject.c#L1740